### PR TITLE
Implement execution environment

### LINKED ISSUE #500: Implement execution environment, both during prepare and aggregate phase
This will involve both designing the data structure and make sure it implements proper interface

### PR DESCRIPTION
**Pull request by taobun**
fixed #500

### CODE REVIEW

File: chain/x/zoracle/exec_env.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+func (env *ExecutionEnvironment) GetAggregateBlockTime() int64 {
+	if int64(len(env.request.ReceivedValidators)) >= env.request.SufficientValidatorCount {
+		return env.ctx.BlockTime().Unix()
+	}
+	return 0
+}
+
+func (env *ExecutionEnvironment) GetValidatorAddress(validatorIndex int64) ([]byte, error) {
+	if validatorIndex >= int64(len(env.request.RequestedValidators)) {
```
Comment (sorawit) [inline chain/x/zoracle/exec_env.go]: What happens if we get a negative number?

Comment (sorawit) [reply chain/x/zoracle/exec_env.go]: Same question for the two functions below.

File: chain/x/zoracle/exec_env.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	return env.request.RequestedValidators[validatorIndex], nil
+}
+
+func (env *ExecutionEnvironment) RequestExternalData(
+	dataSourceID int64,
+	externalDataID int64,
+	calldata []byte,
+) error {
+	env.keeper.SetRawDataRequest(env.ctx, env.requestID, externalDataID, dataSourceID, calldata)
```
Comment (sorawit) [inline chain/x/zoracle/exec_env.go]: This must fail if the dataSourceID does not exist.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/exec_env.go (hunk 2 above)
Comment (sorawit) [reply chain/x/zoracle/exec_env.go]: Must also fail if externalData is already used

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/exec_env.go (hunk 2 above)
Comment (sorawit) [reply chain/x/zoracle/exec_env.go]: Must also fail if the calldata size is too big. Pending on #531

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/exec_env.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+func (env *ExecutionEnvironment) GetExternalData(
+	externalDataID int64,
+	validatorIndex int64,
+) ([]byte, error) {
+	if validatorIndex < 0 || validatorIndex >= int64(len(env.request.RequestedValidators)) {
+		return []byte{}, errors.New("validator out of range")
+	}
+	validatorAddress := env.request.RequestedValidators[validatorIndex]
+
+	report, err := env.keeper.GetRawDataReport(env.ctx, env.requestID, externalDataID, validatorAddress)
```
Comment (sorawit) [inline chain/x/zoracle/exec_env.go]: Why don't we just `return` here?

File: chain/x/zoracle/internal/keeper/raw_request.go [hunk 4]
```diff
@@ -28,3 +47,8 @@ func (k Keeper) GetRawDataRequest(
 	k.cdc.MustUnmarshalBinaryBare(bz, &requestDetail)
 	return requestDetail, nil
 }
+
+func (k Keeper) CheckRawDataRequestExists(ctx sdk.Context, requestID, externalID int64) bool {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/raw_request.go:53]: Comment?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/exec_env.go (hunk 3 above)
Comment (taobun) [reply chain/x/zoracle/exec_env.go]: Good catch!

**Review by sorawit — APPROVED** [review]
Looks good

**Review by evilpeach — APPROVED** [review]

**Pull request merged**